### ISSUE TITLE
NuGet restore memory allocations causes frequent GC which leads to UI delays and slow performance

### ISSUE THREAD

**Issue by Pilchie** [NuGet/Home#1335]
Some large projects (cough, Roslyn, cough) may want to disable restore on build support because restore takes a long time.  However, it would be nice to have that be something configured by the project, instead of it being a per-user setting that we have to tell everyone who clones our solution to go change.

**Comment by davidfowl** [comment]
Just going to leave this right here :smile:

**Comment by nkolev92** [comment]
Thanks for filing this issue. 
We believe this issue is now redundant considering the direction where NuGet/.NET eco-system has gone with an emphasis on performance

**Comment by Pilchie** [comment]
Even with recent improvements in 15.6, most Roslyn devs still leave this option disabled, as it still takes 10s of seconds to restore.  Tagging @jaredpar

**Comment by Pilchie** [comment]
@rrelyea @emgarten NOTE - in #2685, many Roslyn devs still feel like they can't work if VS has the option to restore packages enabled.  If this isn't going to be something configurable for a repo, we need someone to investigate the perf problems the team is seeing.

**Comment by Pilchie** [comment]
See also https://github.com/dotnet/roslyn/pull/24411, which seems to imply that it works?

**Comment by nkolev92** [comment]
@Pilchie We definitely would prefer to solve the perf issues. 

As Justin suggested in the other issue, there are some improvements coming in 15.6 as well. 

I looked at the config, and I don't think that ever worked. 
We don't read the automatic flag.

**Comment by jainaashish** [comment]
`We don't read the automatic flag.` 
This isn't correct, we read this flag and it's per user settings. 

@Pilchie Are you still talking about build restore or auto restore? we have a worktime to specifically disable auto restore which is not available as of now. It won't be easy to have this setting per project, since it's tied with existing NuGet options which have always been per user settings. 

Other than that, I'd imagine most times, build restore will be noop so it should be pretty quick. So can you provide more details for your specific cases like:
- Do you see NuGet running actual restore more often then required?
- Does NoOp also takes 10s of seconds to complete?
- Please capture ETL traces whenever you see it's taking longer than expected and we'll further investigate.

**Comment by sharwell** [comment]
> Do you see NuGet running actual restore more often then required?

Yes. We run a command line restore before starting Visual Studio to avoid the problems caused by a restore inside the IDE. This sequence does not work, and the restore happens inside the IDE anyway.

> Does NoOp also takes 10s of seconds to complete?

We have no way to know what is NOP and what is not. Our response is limited to cases where we believe it is logical to conclude that the restore *should* be a NOP. On this, the answer is certainly yes, pointing back to my previous answer.

> Please capture ETL traces whenever you see it's taking longer than expected and we'll further investigate.

I would recommend using the **General Purpose** and/or **GC Only** scenarios described on [this page](https://github.com/dotnet/roslyn/wiki/Recording-performance-traces-with-PerfView) to capture the following sequence:

1. Clone dotnet/roslyn
2. Run **git clean -dxf**
3. Run **Restore.cmd**
4. Remove [this line](https://github.com/dotnet/roslyn/blob/2aa793ee7b1ff7d3fc944ab6b99679c5664eac94/NuGet.Config#L8) which is currently disabling NuGet functionality inside the IDE
4. Open the IDE and wait for full semantic highlighting to be present on [RoslynPackage.cs](https://github.com/dotnet/roslyn/blob/master/src/VisualStudio/Core/Def/RoslynPackage.cs)

The majority of delays experienced during the test are caused by GC pauses. The majority of allocations (which lead to GC pauses) are within NuGet code.

**Comment by nugetlurker** [comment]
`Yes. We run a command line restore before starting Visual Studio to avoid the problems caused by a restore inside the IDE. This sequence does not work, and the restore happens inside the IDE anyway`

We've recently fix few issues here to make restore consistent across CLI and VS which are part of 15.6 preview4 release. So things should certainly be improved here. But we'll continue to analyze it further for the allocation part. Thanks for the pointers!

**Comment by jainaashish** [comment]
Yes. We run a command line restore before starting Visual Studio to avoid the problems caused by a restore inside the IDE. This sequence does not work, and the restore happens inside the IDE anyway
We've recently fix few issues here to make restore consistent across CLI and VS which are part of 15.6 preview4 release. So things should certainly be improved here. But we'll continue to analyze it further for the allocation part. Thanks for the pointers!

I'll reopen the bug and change the title

**Comment by nkolev92** [comment]
When the Build logs are set to detailed, NuGet prints a log statement that indicates whether it did a full restore or a noop. Just for future reference. 
Example:
```
Restoring NuGet packages...
Running restore with 8 concurrent jobs.
Reading project file C:\Users\nikolev.REDMOND\Source\Repos\ConsoleApp22\ConsoleApp22\ConsoleApp22.csproj.
The restore inputs for 'ConsoleApp22' have not changed. No further actions are required to complete the restore.
Committing restore...
Assets file has not changed. Skipping assets file writing. Path: C:\Users\nikolev.REDMOND\Source\Repos\ConsoleApp22\ConsoleApp22\obj\project.assets.json
No-Op restore. The cache will not be updated. Path: C:\Users\nikolev.REDMOND\Source\Repos\ConsoleApp22\ConsoleApp22\obj\ConsoleApp22.csproj.nuget.cache
Restore completed in 1.89 ms for C:\Users\nikolev.REDMOND\Source\Repos\ConsoleApp22\ConsoleApp22\ConsoleApp22.csproj.
All packages are already installed and there is nothing to restore.
Time Elapsed: 00:00:00.0165513
```

**Comment by aortiz-msft** [comment]
Changing title to reflect the original comment.

https://docs.microsoft.com/en-us/nuget/reference/nuget-config-file#packagerestore-section has an example on how to disable automatic restore in Nuget.config file.

**Comment by aortiz-msft** [comment]
Please feel free to create new issues as needed.
